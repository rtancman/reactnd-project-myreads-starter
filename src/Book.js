import React, { Component } from 'react';
import PropTypes from 'prop-types'


const shelfOptions = [
  {value: 'currentlyReading', label: 'Currently Reading'},
  {value: 'wantToRead', label: 'Want to Read'},
  {value: 'read', label: 'Read'},
  {value: 'none', label: 'None'}
]

class Book extends Component {
  static propTypes = {
    book: PropTypes.object.isRequired,
    onMoveShelf: PropTypes.func.isRequired
  }

  state = {
    book: this.props.book
  }

  changeShelf = (shelf) => {
    let current = this.state.book
    this.props.onMoveShelf(current, shelf)
    current.shelf = shelf
    this.setState({book: current})
  }

  render() {
    const { book } = this.props
    return (
      <div className="book">
        <div className="book-top">
          {book.imageLinks && book.imageLinks.thumbnail && (
            <div className="book-cover" style={{ width: 128, height: 193, backgroundImage: `url("${book.imageLinks.thumbnail}")` }}></div>
          )}
          <div className="book-shelf-changer">
            <select value={this.state.book.shelf} onChange={(event) => this.changeShelf(event.target.value)} >
              <option value="move" disabled>Move to...</option>
              {shelfOptions.map((shelf) => (
                <option key={shelf.value} defaultValue={this.state.book.shelf === shelf.value} value={shelf.value}>{shelf.label}</option>
              ))}
            </select>
          </div>
        </div>
        <div className="book-title">{book.title}</div>
        {book.authors && (
          <div className="book-authors">{book.authors.join(', ')}</div>
        )}
      </div>
    )
  }
}

export default Book

