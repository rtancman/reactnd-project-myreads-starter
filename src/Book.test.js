import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount, render } from 'enzyme';
import Book from './Book'
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom'

const book = {
  'title': 'The Linux Command Line',
  'subtitle': 'A Complete Introduction',
  'authors': [
      'William E. Shotts, Jr.'
  ],
  'publisher': 'No Starch Press',
  'publishedDate': '2012',
  'description': 'You have experienced the shiny, point-and-click surface of your Linux computerâ€”now dive below and explore its depths with the power of the command line. The Linux Command Line takes you from your very first terminal keystrokes to writing full programs in Bash, the most popular Linux shell.',
  'industryIdentifiers': [
      {
          'type': 'ISBN_13',
          'identifier': '9781593273897'
      },
      {
          'type': 'ISBN_10',
          'identifier': '1593273894'
      }
  ],
  'readingModes': {
      'text': true,
      'image': false
  },
  'pageCount': 480,
  'printType': 'BOOK',
  'categories': [
      'COMPUTERS'
  ],
  'averageRating': 4,
  'ratingsCount': 2,
  'maturityRating': 'NOT_MATURE',
  'allowAnonLogging': true,
  'contentVersion': '1.2.2.0.preview.2',
  'panelizationSummary': {
      'containsEpubBubbles': false,
      'containsImageBubbles': false
  },
  'imageLinks': {
      'smallThumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api',
      'thumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'
  },
  'language': 'en',
  'previewLink': 'http://books.google.com/books?id=nggnmAEACAAJ&dq=linux&hl=&cd=3&source=gbs_api',
  'infoLink': 'https://play.google.com/store/books/details?id=nggnmAEACAAJ&source=gbs_api',
  'canonicalVolumeLink': 'https://market.android.com/details?id=book-nggnmAEACAAJ',
  'id': 'nggnmAEACAAJ',
  'shelf': 'currentlyReading'
}

const onMoveShelf = jest.fn()

describe('Book', () => {
  describe('when component renders', () => {

    it('without crashing', () => {
      const component = renderer.create(
        <Book book={book} onMoveShelf={onMoveShelf} />,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('without imageLinks', () => {
      const book = {
        'title': 'lala',
        'authors': [
            'William E. Shotts, Jr.'
        ]
      }
      const component = renderer.create(
        <Book book={book} onMoveShelf={onMoveShelf} />,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('without authors', () => {
      const book = {
        'title': 'lala',
        'imageLinks': {
            'smallThumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=5&source=gbs_api',
            'thumbnail': 'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api'
        }
      }
      const component = renderer.create(
        <Book book={book} onMoveShelf={onMoveShelf} />,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('when user change category book', () => {
    it('change select value', () => {
      const onMoveShelf = jest.fn(() => {
        return new Promise((resolve, reject) => resolve())
      })
      const event = {
        preventDefault() {},
        target: { value: 'read' }
      };
      const component = shallow(<Book book={book} onMoveShelf={onMoveShelf} />);
      component.find('select').simulate('change', event);
      expect(onMoveShelf).toBeCalled();
      expect(component.state().book.shelf).toEqual('read');
    });
  });
});

