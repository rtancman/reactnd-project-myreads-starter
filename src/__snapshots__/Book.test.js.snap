// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book when component renders without authors 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(\\"http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api\\")",
          "height": 193,
          "width": 128,
        }
      }
    />
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value={undefined}
      >
        <option
          disabled={true}
          value="move"
        >
          Move to...
        </option>
        <option
          defaultValue={false}
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          defaultValue={false}
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          defaultValue={false}
          value="read"
        >
          Read
        </option>
        <option
          defaultValue={false}
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  >
    lala
  </div>
</div>
`;

exports[`Book when component renders without crashing 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(\\"http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api\\")",
          "height": 193,
          "width": 128,
        }
      }
    />
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value="currentlyReading"
      >
        <option
          disabled={true}
          value="move"
        >
          Move to...
        </option>
        <option
          defaultValue={true}
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          defaultValue={false}
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          defaultValue={false}
          value="read"
        >
          Read
        </option>
        <option
          defaultValue={false}
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  >
    The Linux Command Line
  </div>
  <div
    className="book-authors"
  >
    William E. Shotts, Jr.
  </div>
</div>
`;

exports[`Book when component renders without imageLinks 1`] = `
<div
  className="book"
>
  <div
    className="book-top"
  >
    <div
      className="book-shelf-changer"
    >
      <select
        onChange={[Function]}
        value={undefined}
      >
        <option
          disabled={true}
          value="move"
        >
          Move to...
        </option>
        <option
          defaultValue={false}
          value="currentlyReading"
        >
          Currently Reading
        </option>
        <option
          defaultValue={false}
          value="wantToRead"
        >
          Want to Read
        </option>
        <option
          defaultValue={false}
          value="read"
        >
          Read
        </option>
        <option
          defaultValue={false}
          value="none"
        >
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title"
  >
    lala
  </div>
  <div
    className="book-authors"
  >
    William E. Shotts, Jr.
  </div>
</div>
`;
